### PR TITLE
Fix `declaration-property-value-no-unknown` false negatives

### PR DESCRIPTION
**Pull request by Mouvedia**
> Which issue, if any, is this issue related to?

follow-up of #7944

> Is there anything in the PR that needs further explanation?

[demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdilTNgAsIAnAWkyQFsZ4BuLHANwEsYB3UgF2eoBtnMMUkgAezAM7wUAMwgR62FC3ZceMfoNIDxMTlNnyGS1h258BQitTpHlptRpoy5Cpia7E845twiZyVE4GCgC+WCAANCDSzLwwAHKBCCAwIlQADnEAdGDi4pHgfjEA5skYiuggxACuceKVUuU4OJWwYLxInr7+6cQQ6TDEnACepIxIvNWWEKTVmADWmBBsmA0onDUwRmGYIQWQmCUAYiSUSJzJAFbifgWw6fmITSiV4iNxGpxrlR2cMG+VCJGV7vdQCTikA4lUhvPBQTpQb4gX7-L4gLB7EJAA)

https://caniuse.com/mdn-css_properties_view-transition-name
https://caniuse.com/?search=view-timeline
https://caniuse.com/mdn-css_properties_anchor-name
https://csstree.github.io/docs/validator.html

The changeset of #7944 will be modified in #7946 to include #7957.
see https://github.com/stylelint/stylelint/pull/7946/files#r1730474501

### CODE REVIEW

File: lib/rules/declaration-property-value-no-unknown/index.mjs [hunk 1]
```diff
@@ -64,16 +64,30 @@ const rule = (primary, secondaryOptions) => {
...
 			'text-wrap-style': 'auto | balance | pretty | stable',
+			'view-timeline-axis': '[ block | inline | x | y ]#',
+			'view-timeline-inset': '[ [ auto | <length-percentage> ]{1,2} ]#',
+			'view-timeline-name': '[ none | <custom-property-name> ]#',
+			'view-timeline':
+				"[ <'view-timeline-name'> [ <'view-timeline-axis'> || <'view-timeline-inset'> ]? ]#",
+			// <custom-ident> represents any valid CSS identifier that would not be misinterpreted as a pre-defined keyword in that property’s value definition
+			// i.e. reserved keywords don't have to excluded
+			// w3c/csswg-drafts#9895
+			'view-transition-name': 'none | <custom-ident>',
```
Comment (Mouvedia) [inline lib/rules/declaration-property-value-no-unknown/index.mjs:82]: I couldn't find any way to exclude a keyword.
i.e. something that could merge `<custom-ident>` with `auto#{0}`

File: lib/rules/declaration-property-value-no-unknown/index.mjs [hunk 2]
```diff
@@ -64,16 +64,30 @@ const rule = (primary, secondaryOptions) => {
 		const propertiesSyntax = {
 			overflow: '| overlay', // csstree/csstree#248
 			width: '| min-intrinsic | -moz-min-content | -moz-available | -webkit-fill-available', // csstree/csstree#242
+			'anchor-name': 'none | <custom-property-name>#',
```
Comment (Mouvedia) [inline lib/rules/declaration-property-value-no-unknown/index.mjs:67]: `<dashed-ident>` is not exposed.

see https://github.com/csstree/csstree/blob/ba6dfd8bb0e33055c05f13803d04825d98dd2d8d/lib/lexer/generic.js#L225-L227

File: lib/rules/declaration-property-value-no-unknown/index.cjs [hunk 3]
```diff
@@ -67,16 +67,30 @@ const rule = (primary, secondaryOptions) => {
...
 			'text-box-trim': 'none | trim-start | trim-end | trim-both',
 			'text-spacing-trim': 'normal | space-all | space-first | trim-start',
 			'text-wrap-style': 'auto | balance | pretty | stable',
+			'view-timeline-axis': '[ block | inline | x | y ]#',
+			'view-timeline-inset': '[ [ auto | <length-percentage> ]{1,2} ]#',
+			'view-timeline-name': '[ none | <custom-property-name> ]#',
+			'view-timeline':
+				"[ <'view-timeline-name'> [ <'view-timeline-axis'> || <'view-timeline-inset'> ]? ]#",
+			// <custom-ident> represents any valid CSS identifier that would not be misinterpreted as a pre-defined keyword in that property’s value definition
+			// i.e. reserved keywords don't have to excluded
```
Comment (Mouvedia) [inline lib/rules/declaration-property-value-no-unknown/index.cjs:83]:
```suggestion
			// i.e. reserved keywords don't have to excluded explicitly
```

**Review by romainmenke — COMMENTED** [review]
I'd rather we fix false negatives upstream in `css-tree` and try to limit local patches to false positives.

I am concerned that we will end up maintaining to much data/config here in Stylelint.

----

Edit: I've asked in `css-tree` about the maintenance status of `csstree` : https://github.com/csstree/csstree/issues/279

**Comment by Mouvedia** [comment]
> I'd rather we fix false negatives upstream in css-tree and try to limit local patches to false positives.

Me too but for now this will do.
I value the users more than _temporary_ annoyances for maintainers.
i.e. we will eventually fork its repo if the license let's us
I also waited years before finally taking the time to send a PR.
It's not ideal but it's the best we can do for _now_.

Good luck with your issue nonetheless.

**Comment by romainmenke** [comment]
My concern is exactly because this is not a temporary situation.
It would be a permanent increase in our workload.

I'd appreciate it if we can open this up for feedback and other opinions and do not immediately assume that this is something that we need to do.

Keep in mind that even if you are willing to submit all the patches to keep this rule up to date, someone else also needs to volunteer their time to review and release :)

**Comment by Mouvedia** [comment]
We have accepted to add the rule on the premise that css-tree would be maintained or accept our PRs in a timely manner.
It wasn't the case.
This rule had accumulated false positives/negatives.
I went and fixed some of them.

> and do not immediately assume that this is something that we need to do.

If there's a false negative/positive, it's an immediate _ready to implement_ unless a browser doesn't support it.
i.e. it's up to be discussed if it's behind a flag or if it has unwanted consequences
i.e. #7510
e.g. #7917

If you don't to maintain this rule anymore, you can open a new issue to discuss its removal/deprecation.
For now Ill continue fixing false negatives/positives for non-deprecated rules.

**Comment by ybiquitous** [comment]
I appreciate @Mouvedia's work to fix bugs in this rule, but I well understand @romainmenke's concern. Especially, this rule has to support many property value definitions to be *truly useful*. We want to rely on a reliable library like csstree to work on this rule, but *for now*, we have to accept ad-hoc fixes like this PR.

Of course, I believe this rule should be useful for many CSS authors because of its error detection.

If too many bug fixes would come for this rule or we would feel overloaded, let's reconsider making it deprecated or stopping the maintenance until csstree becomes active or an alternative appears.

Generally speaking, this is an example of challenging things in OSS management. 😞

**Comment by Mouvedia** [comment]
> If too many bug fixes would come for this rule or we would feel overloaded, let's reconsider making it deprecated or stopping the maintenance until csstree becomes active or an alternative appears.

If it comes to that, we will add a new `stylelintWarningType`.
e.g. `frozen`

If we get no answers from css-tree maintainers, I would prefer if we'd go the fork route.
Well let's not get ahead of ourselves; the options are almost enough for now.
I am afraid we won't be able to fix #7241 in a spec-compliant manner through them though.

**Review by ybiquitous — COMMENTED** [review]
The changes look good to me. 👍🏼 

Can you add a changelog item? If possible, I want to clarify the target properties by this bug fix, e.g.

```
Fix `declaration-property-value-no-unknown` false negatives for <property names...>
```

**Comment by Mouvedia** [comment]
> The changes look good to me. 👍🏼
> 
> Can you add a changelog item? If possible, I want to clarify the target properties by this bug fix, e.g.
> 
> ```
> Fix `declaration-property-value-no-unknown` false negatives for <property names...>
> ```

> The changeset of https://github.com/stylelint/stylelint/pull/7944 will be modified in https://github.com/stylelint/stylelint/pull/7946 to include https://github.com/stylelint/stylelint/pull/7957.
> see [#7946 (files)](https://github.com/stylelint/stylelint/pull/7946/files#r1730474501)

**Review by ybiquitous — APPROVED** [review]
Okay, let's manually add this PR in #7946. 👍🏼

**Comment by romainmenke** [comment]
I think we are overlooking a third option :)

`csstree` is extensible, exactly this is what allows us to patch it here in the first place.

That I think that Stylelint is not the right context doesn't mean that I don't want to have a more modern syntax definition.

I've already referred to `@webref/css` multiple times and I think it also offers a solution here.

------

My main concern with these patches:
- they are adhoc
- they are manual work
- they are not fully and exhaustively tested (we don't even call and check `lexer.validate()`)
- they are not shareable

----

I think a better approach is to have a shared library with syntax patches.

A library that uses external references like `@webref/css` to automated most of the work.
A library that is separate from Stylelint so that it can be used by others facing the same issue.

Something roughly like this: https://github.com/csstools/postcss-plugins/pull/1466

I am not looking forward to maintaining these syntax definitions by hand, here in Stylelint.
But I don't mind maintaining a generated and fully test definition in a context that can support such an effort :)

I'll try to allocate some time to this over the next few weeks and open an issue if I think it is ready enough for feedback.

**Comment by ybiquitous** [comment]
I'm so curious about the work of [`@webref/css`](https://www.npmjs.com/package/@webref/css), thanks for the info, @romainmenke. 😄 

Let me ask a question, isn't there a smarter way to replace csstree by using `@webref/css` for now? I'm concerned about whether we will have to care about csstree.

**Comment by Mouvedia** [comment]
@romainmenke I am curious to know how you handled [this problem](https://github.com/stylelint/stylelint/issues/7241#issuecomment-2302913135).
i.e. I was under the impression that it would require to update css-tree itself

**Comment by romainmenke** [comment]
What `csstree` got right was being able to parse the value definitions (e.g. `'alpha-value': '<number> | <percentage>'`)

This works really well and CSS value syntax evolves slowly enough that this will continue to work. We can still consider forking `csstree` if there are sufficient future issues with this aspect.

At this time I don't see an urgent need to make updates to `csstree` internals.

The more immediate problem is that `csstree` isn't publishing regular updates with the latest syntax definitions.

I propose not doing these syntax definition updates, here, by hand. But instead do them automated and shareable, separate from Stylelint.

These patches are just a mapping of properties and types types and their respective syntax.
`@webref/css` already contains all this information and it is very actively maintained by the W3C.

This approach is also relatively safe going forwards:
- the current script I am writing detects if `csstree` has an update and will request updated patches for those entries. So when `csstree` is more actively maintained the list of patches will shrink.
- the list of patches is just a json and can be copy/pasted and manually maintained if I stop doing maintenance.

I am just a very big proponent of not doing manual and error prone work when even remotely possible 😛

------

> I guess they can maybe be added individually.

Indeed, that is the way to go :)

**Comment by Mouvedia** [comment]
Diversities in approaches is a boon for OSS projects.
If I didn't act and patched it, it wouldn't have tickled you enough.
This was worth it after all, let's switch to what you are proposing once it's ready.

**Comment by ybiquitous** [comment]
I understand it well; thanks for the explanation. 😃 

I'm looking forward to the published list of patches to `csstree` based on `@webref/css`. Then, let's replace the current implementation of the new package. 👍🏼

**Comment by romainmenke** [comment]
> If I didn't act and patched it, it wouldn't have tickled you enough.

No, I am already working 10-12 hours a day and rarely take a day off 😓

By patching these here you made it more urgent do something to illustrate a healthier way forwards. I was well aware of how this could be improved and it was on my very long todo list. I could have proactively communicated this, but that also takes time, that is on me. 

Making changes that result in even more work is alarming to me exactly because of the already massive workload. Quickly showing how we can steer this back to less maintenance work is taxing but important to make sure things don't escalate.

The only way I see that OSS maintenance can be done in a sustainable way is by leveraging automations and minimizing manual labor. This type of manual labor just doesn't scales in OSS. We can never keep up with the collective effort of all spec editors and implementers working for Apple, Mozilla, Google, Samsung, Microsoft, ...

Again this doesn't mean that I don't appreciate these improvements in PR's like this 🙇 

But be mindful that there is a cost to all this.
I have to carefully choose where I spend my time to make sure I don't burn myself out.

**Comment by Mouvedia** [comment]
> > If too many bug fixes would come for this rule or we would feel overloaded, let's reconsider making it deprecated or stopping the maintenance until csstree becomes active or an alternative appears.
> 
> If it comes to that, we will add a new `stylelintWarningType`. e.g. `frozen`

related: https://github.com/eslint/rewrite/pull/149/files#diff-4c2458380e50c967c70d7afe819dc7412180f9f817043f27441de3498e4d0479R136
i.e. `frozen` would be boolean property of `RuleMeta`

**Pull request merged**